core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -715,6 +715,20 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
   def tap[R1 <: R, E1 >: E](f: A => ZManaged[R1, E1, Any]): ZManaged[R1, E1, A] =
     flatMap(a => f(a).as(a))
 
+  /**
+   * Returns an effect that effectfully peeks at the failure or success of the acquired resource.
+   */
+  final def tapBoth[R1 <: R, E1 >: E](f: E => ZManaged[R1, E1, Any])(g: A => ZManaged[R1, E1, Any])(
+    implicit ev: CanFail[E] // curried for better type inference
+  ): ZManaged[R1, E1, A] =
+    flatMap(a => g(a) *> tapError(f))
+
+  /**
+   * Returns an effect that effectfully peeks at the failure of the acquired resource.
+   */
+  final def tapError[R1 <: R, E1 >: E](f: E => ZManaged[R1, E1, Any])(implicit ev: CanFail[E]): ZManaged[R1, E1, A] =
+    flatMap(a => flip.tap(f).flip.as(a))
```

Comment:
Pretty sure you need to implement this using `foldM`, otherwise `f` will not be executed when `this` fails.